README

INSTALLATION : 
   See the file named INSTALL in the top directory 
  
STANDARD USAGE : 
   spice [-keyword(1) parameter(1) ... -keyword(N) parameter(N)] 
         [-optinfile parameter_file]
  
   All keywords are optional
  
 Format of parameter_file :
              keyword = parameter
       see example below

LIST OF KEYWORDS :     keyword [possible parameters](default)

-about
   print copyright info and stop

-alm1fileout [name|YES|NO](NO)
   FITS file containing a_lm coefficients of (masked and weighted) map1
   NO desactivates output of the alm
   YES activates output of the alm into alm1.fits

-alm2fileout [name|YES|NO](NO)
   FITS file containing a_lm coefficients of (masked and weighted) map2
   NO desactivates output of the alm
   YES activates output of the alm into alm1.fits

-apodizesigma [dfloat|NO](NO)
   scale factor in DEGREES of the correlation function tappering,
   see option -apodizetype.
   For better results, -apodizesigma should be close to -thetamax.
   NO desactivates apodization.
   NOTE : if noise subtraction is active (options -noisecorfile or
   -noiseclfile), the input noise data are assumed to be already apodized.

-apodizetype [int](0)
   type of apodization
   0: the correlation function is multiplied by a gaussian window
      (equal to 1    at theta=0,
       equal to 0.5  at theta= -apodizesigma/2,
       equal to 1/16 at theta= -apodizesigma).
   1: the correlation function is multiplied by a cosine window 
      (equal to 1   at theta=0,
       equal to 0.5 at theta= -apodizesigma/2,
       equal to 0   at theta= -apodizesigma).

-beam [dfloat|NO](NO)
   gaussian beam FWHM (arcmin) for map 1.
   If specified, output c_l and c_theta are corrected for it
   NOTE : if noise subtraction is active (options -noisecorfile or
   -noiseclfile), the input noise data are assumed to be already corrected
   for the beam

-beam_file [name|NO](NO)
   beam window function for map 1.
   If specified, output c_l and c_theta are corrected for it
   The file should be plain ASCII with 2 columns : l and B(l) 
   or FITS (ASCII or BINARY) table containing B(l) starting at l=0
   (B(l) is sqrt of beam power spectrum).
   This over-rules the gaussian correction described by beam.
   NO will not read any beam_file (but a gaussian
      correction may be performed if -beam is set).
   NOTE : if noise subtraction is active (options -noisecorfile or
   -noiseclfile), the input noise data are assumed to be already corrected
   for the beam.
   Starting with version 2.90 and if linked with HEALPix 3.10+
   file.fits[extname] means extension 'extname' of file.fits

-beam2 [dfloat|NO](NO)
   gaussian beam FWHM (arcmin) for map 2 (if applicable).
   (see beam)
  Note: if the second beam is the same as the first one,
   then -beam2 should be explicitely given the same value as -beam

-beam_file2 [name|NO](NO)
   beam window function for map 2 (if applicable).
   If specified, output c_l and c_theta are corrected for it
   The file should be plain ASCII with 2 columns : l and B(l) 
   or FITS (ASCII or BINARY) table containing B(l) starting at l=0
   (B(l) is sqrt of beam power spectrum).
   This over-rules the gaussian correction described by -beam2.
   NO will not read any -beam_file2 (but a gaussian
      correction may be performed if -beam2 is set.
   NOTE : if noise subtraction is active (options -noisecorfile or
   -noiseclfile), the input noise data are assumed to be already corrected
   for the beam.
  Note: if the second beam is the same as the first one,
   then -beam_file2 should be explicitely given the same value as -beam_file

-clfile [name|YES|NO](NO)
   name of output file for Cl
   YES activates Cl file output with name spice.cl
   NO desactivates output of Cl file
   NOTE : the current format of this file is a two columns array
   the first one corresponding to the value of l, the second one
   to the temperature Cl and, if polarization is activated, five
   additional spectra corresponding to EE, BB, TE, TB, EB 
   and optionally ET, BT, BE are also written.

-cl_inmap_file [name|YES|NO](NO)
   name of input file containing the raw Cls of the map as obtained by
   running previously SpICE with the option -cl_outmap_file.
   If this option is active, it is not necessary anymore to read the input
   map file, and option -cl_outmap_file is of course inactive.
   YES reads the map raw Cls from file spice.clrawmap
   NO computes the raw Cl directly from the map.
   IMPORTANT NOTE : the cls computed that way can be affected by the nature
      of the masks and/or the weights. If the masks/weights are changed,
      the calculation will become inconsistent.

-cl_inmask_file [name|YES|NO](NO)
   name of input file containing the raw Cl of the weight/masks as
   obtained by running previously SpICE with the option -cl_outmask_file.
   If option -cl_inmask_file is active, the Cls of the masks are not computed,
   they are directly read in file specified by cl_inmask_file and option
   -cl_outmask_file is of course inactive.
   With option -cl_inmask_file active, it is in principle not necessary
   to read masks nor weights, but if masks/weights are not read, it means
   that the input map is supposed to have been processed prior to using spice,
   i.e. it has been multiplied by the masks/weights already.
   YES reads the weights/mask raw Cls from file spice.clrawmasks
   NO computes the raw Cl directly from the weights/masks.
   IMPORTANT NOTE : the cls computed for the map are affected by the nature
      of the masks and/or the weights. If the masks/weights are changed,
      the calculation will become inconsistent.

-cl_outmap_file [name|YES|NO](NO)
   name of output file containing the raw Cls of the map. This can be
   convenient in order to run SpICE with different parameters, without
   having to recompute these raw Cls.
   YES activates map raw Cls output with name spice.clrawmap
   NO desactivates output of map raw Cls file.

-cl_outmask_file [name|YES|NO](NO)
   name of output file containing the raw Cls of the weights/masks.
   This can be convenient in order to run SpICE with different parameters,
   without having to recompute these raw Cls.
   YES activates mask raw Cls output with name spice.clrawmasks
   NO desactivates output of mask raw Cls file.

-corfile [name|YES|NO](NO)
   name of output correlation file
   YES activates correlation file output with name spice.cor
   NO desactivates output of correlation file
   NOTE : the current format of this file is a three columns array
   the first one corresponding to the separation theta in radians
   the second one to cos(theta), the third one to the correlation
   function of the temperature and, if polarization is activated,
   five additional columns are written, corresponding respectively to
   QQ, UU, TQ, TU and QU in the coupled case and to EE, BB, TQ, TU and QU
   in the decoupled case (see option -decouple and -symmetric_cl).

-covfileout [name|YES|NO](NO)
   Computes and ouputs the TT C(l) covariance matrices 
     These matrices are not needed by the code and are provided for 
   eg, cosmological interpretation of the result.
   YES activates output of the cov. matrices in default
      FITS file spice.covariance
   NO desactivates output of the matrices

-decouple [YES|NO](NO)
   computes the decoupled correlation functions
   by an integral over QQ and UU correlation functions in angle space.
   NOTE : this option is active only if -polarization=YES,
   otherwise it is ignored (see also -tolerance).
   YES activates decoupling :
      Correlation functions computed are : TT, EE, BB, TQ, TU, QU, (QT, UT, UQ)
      Cls computed are : TT, EE, BB, TE, TB, EB, (ET, BT, BE) 
      with no mixing between EE and BB
   NO desactivates decoupling :
      Correlation functions computed are : TT, QQ, UU, TQ, TU, QU, (QT, UT, UQ)
      Cls computed are : TT, EE, BB, TE, TB, EB, (ET, BT, BE) 
      with coupling between EE and BB
 
    In both cases, the TE (and TB) estimators are unbiased using window functions (see
      options -tenormfilein, -tenormfileout, -windowfilein, -windowfileout).

-dry [YES|NO](NO)
   This unsafe option allows reading partly corrupted files, i.e. with
   incomplete headers. However the size of the fits file must be correct,
   i.e. it must correspond to a full sky map with 12*nside^2 pixels.

-extramapfile [name|NO](NO)
   name of the (optional) extra input map file (healpix SP fits file)
   to be added to mapfile before analysis. 
   Must be in the some Coordinates, Nside and Units as mapfile
   (ordering can be different).

-extramapfile2 [name|NO](NO)
   name of the (optional) extra input map file (healpix SP fits file)
   to be added to mapfile2 before analysis. 
   Must be in the some Coordinates, Nside and Units as mapfile2
   (ordering can be different).

-fits_out [YES|NO](NO)
   If set, the Cl and correlation (-clfile, -corfile, -cl_outmap_file, -cl_outmask_file)
   are written in FITS Ascii files, instead of plain text files.
   This does not affect the input Cl files (-cl_inmap_file and -cl_inmask_file),
   which will be read whatever is their format.

-help
   print this message and stop

-history
   print history of the program and stop

-kernelsfileout [name|YES|NO](NO)
   Compute and ouput the 4 kernels, each of size (lmax+1)*(2*lmax+1),
   relating the Spice (polarized) estimator to the 'true' underlying CMB spectra.
     These kernels are not needed by the code and are provided for 
   eg, cosmological interpretation of the result.
     Note that these kernels depend on the choice
   of apodization (options -apodizetype and -apodizesigma) 
   and on -thetamax and -nlmax.
     How to use them depends on the -decouple option 
   (see the FITS file header for more information).
     This option is inactive if option -windowfilein is active.
   YES activates output of the kernels in default
      FITS file spice.kernels
   NO desactivates output of the kernels

-listmapfiles1_K [name|NO](NO)
   with K integer in [1, 10]
   list of map files to be combined with weights -listmapweights1_K
   to form the first map to analyze.
   If this option is set, -mapfile (and -extramapfile) are ignored

-listmapfiles2_K [name|NO](NO)
   with K integer in [1, 10]
   list of map files to be combined with weights -listmapweights2_K
   to form the optional second map to cross-correlate with the first one.
   If this option is set, -mapfile2 (and -extramapfile2) are ignored

-listmapweights1_K [dfloat|NO](1.0)
   with K integer in [1, 10]
   list of weights to be applied to -listmapfiles1_K above
   to form the first map to analyze.
   m1 = w1(1) * m1(1) + w1(2) * m1(2) + ... 

-listmapweights2_K [dfloat|NO](1.0)

HISTORY
HISTORY :
+++++++++

UNPOLARIZED SPICE :

07/01/01/SP&IS       : version 0.0
                       - Basic version of the code, xi and Cl
07/15/01/SC&SP&IS    : version 0.1
                       - Improve input/output
                       - Build friendly interface
                       - pixel window and beam window correction
11/16/01/SC&SP       : version 0.2
                       - add pair count thresholding, option
                         npairsthreshold
                       - add apodizationg of the 2pt correlation
                         function, option apodizesigma
                       - add unsafe reading option, option dry
22/05/02/SC&SP       : version 1.0
                       - correct for a bug : the average from
                         the whole map was always subtracted
                         from it. Instead, an option subav
                         is added, where there the (weighted)
                         average is subtracted from the map
                         if required
07/06/02/SC&SP       : version 1.01
                       - correct for a bug in the handling
                         of the flags coroutput, apodize,
                         verbose and megaverbose in subroutine
                         compute.
                       - correct for a bug in the wrapper
                         normfac option was here twice.

POLARIZED SPICE :

01/04/02/SC&EH&SP&IS : version 2.0
                       - add option polarization to compute
                         polarization Cls and correlation functions
07/06/02/EH&SP       : version 2.01
                       - add decouple option to compute pure
                         E and B modes correlation functions
                         and power spectra
07/06/02/SC&SP       : version 2.02
                       - clean up wrapper.
07/03/03/EH          : version 2.1
                       - increase compatibility with Healpix 1.20
                       - introduce nlmax, beam_file, apodizetype
10/15/04/SC&SP       : version 2.2
                       - clean up the code
                       - introduce cl_outmap_file, cl_inmap_file,
                         cl_outmask_file, cl_inmask_file,
                         windowfileout, windowfilein.
                       - PIOLIB compatibility              [Planck]
12/04/06/EH&SP       : version 2.21
                        - bug correction in pointer boundaries
                        - bug correction in PIOLIB output   [Planck]
                        - slight edition of documentation (about noise)
05/10/06/EH          : version 2.22
                        - maps cross-correlation
??      /EH          : version 2.30
                        - different beam for each map
02/02/07/EH          : version 2.40
                        - fixed default behavior of second beam option
                        - reorganized help message
2007-03-08/EH        : version 2.45
                        - first stab at PIE-ization        [Planck]
                        - allow different mask for I and (Q,U)
2007-04-04/EH        : version 2.46
                        - simpler ASCII output format for g95
                        - accepts polarized cut sky maps
                        - crashing bug correction in compute windows
2007-04-18/EH        : version 2.47
                        - some more bug correction (nmask=1 by default)
2007-04-27/EH        : version 2.48
                       - PIE-ization done                 [Planck]
                        +++ Delouis test with thinC        [Planck]
2007-04-27/EH        : version 2.49
                       - bug correction in Legendre for x=-1
2007-09-10/EH        : version 2.50
                       - set bad pixels to 0                [Planck]
                       - added fits_out for FITS C(l) output      
                       - added version option                     
2007-12-07/EH        : version 2.52
                       - added transfer function                  
                       - Healpix version 2.* or more is required  
2007-12-27/EH        : version 2.53
                       - correctly account for DMC beam file [Planck] 
2008-02-29/EH        : version 2.54
                       - copy run parameters in FITS header
2008-04-23/EH        : version 2.55
                       - allow for different weight files and powers 
                        for temp. and pol. maps
2008-05-29/EH        : version 2.560
                       - compute TB and EB spectra
2008-06-02/EH        : version 2.57
                       - correct TB and EB from window functions, update documentation
2008-07-01/EH        : version 2.57a
                       - corrected QU correlation: QU_new = -2 * QU_old
2008-08-25/EH        : version 2.57b
                       - add subdipole option (-> corrects bug on monopole subtraction introduced in v2.55)
2008-09-02/EH        : version 2.59
                       - use DMCPIOSTRINGMAXLEN instead of hardcoded value [Planck]
                       - automatic generation of IO routines with PIE [Planck]
2008-09-11/EH        : version 2.60
                       - compiles on CC-in2p3 [Planck]
2008-10-01/EH        : version 2.61

src/ispice.py
"""
  ispice defines tools to run Spice from Python,
  either in the Planck HFI DMC (aka piolib, objects managed by database)
  or using FITS files
  
Includes:
  Main routine:
        ispice()
  Support routines:
        flatten()
        is_fits()
        look_for_file()
        thinc_exists()
        tools()
        tf2yn()
        subreadme()
        readme()

HISTORY:
     2010-10: first version in python2
     2015-10: introduction in public release
     2019-06: supports python3
     2020-01: supports True/False flags besides "YES"/"NO";
       access to HTML documentation
     2020-05: added support for omp_num_threads, 
         kernelsfileout, noiseclfile, noisecorfile, normfac
     2020-09: added alm?fileout
  """

import os
from os.path import dirname
readmehtml = "http://www2.iap.fr/users/hivon/software/PolSpice/README.html"
ompntdef = os.getenv("OMP_NUM_THREADS","0") # default value of omp_num_threads

def flatten(x):
    """ flatten():  turns nested list or tuple into single list """
    result = []
    if isinstance(x, str):
        result = x
    else:
        for el in x:
            if hasattr(el, "__iter__") and not isinstance(el, str):
                result.extend(flatten(el))
            else:
                result.append(el)
    return result

def is_fits(inlist):
    """ test presence of FITS file in a list """
    import os
    flatlist = flatten(inlist)
    if isinstance(flatlist, str): flatlist = [ flatlist ]
    isFits = False
    for f in flatlist:
        if isinstance(f, str):
            if os.path.exists(f):
                isFits = os.path.isfile(f)
    return isFits

def thinc_exists():
    """ thinc_exists():  test existence of thinc module (DMC specific) """
    import imp
    try:
        imp.find_module('thinc')
        imp.find_module('piolib')
        myDMC = True
    except ImportError:
        myDMC = False
    return myDMC

def look_for_file(lookfor = "spice"):
    """ look for spice code in neighbouring directories """
    import os
    from os.path import join, dirname

    bdir = dirname(__file__)
    # locs = (join(bdir,'.'), join(bdir,'..'), join(bdir,'../..'))
    locs = (join(bdir,'.'), join(bdir,'..'))
    path = ''
    for loc in locs:
        for root, dirs, files in os.walk(loc):
            if lookfor in files:
                path = join(root, lookfor)
                break
        if (path != ''): break
    return path


def tf2yn(arg):
    """ Turn True/False into "YES"/"NO" """
    out = arg
    if (type(arg) == bool):
        out = "YES" if (arg) else "NO"
    return out

def subreadme():
    try:
        webbrowser.open_new_tab(readmehtml)
    except:
        print ("The web browser could not be opened.")
        print ("See %s for PolSpice documentation,"%(readmehtml))
        print ("or type    spice -help  ")
    return

def readme():
    import webbrowser
    
    locrh = look_for_file('README.html')
    if (locrh != ''):
        try:
            webbrowser.open_new_tab("file://%s"%(locrh))
        except:
            subreadme()
    else:
        subreadme()
    return
readme.__doc__="""
    readme() opens a web browser tab with the PolSpice documentation README.html
    found in {0}
    or in    {1}
""".format(dirname(dirname(__file__)),readmehtml)

#----------------------------------------------------------------

class tools(object):
    """ tools() defines context dependent routines """
    def __init__(self, object):
        self.type = object
        in_thinc = thinc_exists()
        in_fits  = is_fits(object)
        if (not in_thinc and not in_fits):
            import sys
            print ('Input files do not exist or are not in FITS'%(object))
            #print ('Input files (%s) do not exist or are not in FITS'%(object))
            #print ('Input files are not in FITS, and DMC not available')
            sys.exit('Aborting')
            
        self.inDMC = in_thinc and not in_fits
        self.command = ''
    # -------------------------------------------
    def myNewJob(self, code, label="", cast=False):
        from os.path import abspath
        if (self.inDMC):
            from thinc import NewJob
            return NewJob(code, label=label, cast=cast)
        else:
            if (code == 'spice' or code == 'spice_SP' or code == 'spice_DP'):
                mycom = look_for_file(code)
            else:
                mycom = code
            self.command += abspath(mycom)+' '
            return 0

    def mySet(self, jobid, name, value_):
        value = tf2yn(value_) # turn True/False into "YES"/"NO"
        if (self.inDMC):
            from thinc import Set
            Set(jobid, name, value)
        else:
            #print str(name)+' = '+str(value)
            self.command += '-%s %s '%(str(name),str(value))

    def mySubmit(self, jobid, exportEnviron=None):
        if (self.inDMC):
            from thinc import Submit
            Submit(jobid, exportEnviron=exportEnviron)
        else:
            import os
            print ('Submitting job %s'%(str(jobid)))
            print (self.command)
            import subprocess
            # Run spice in a shell and print standard output to the Python session
            # `universal_newlines` interprets the output as text and encodes it properly
            # `env` overrides the environment, so we need to get the current environment from `os.environ` and modify it with the exportEnviron dict
            # `stderr.subprocess.STDOUT` makes sure that also stderr is printed
            print(subprocess.check_output(self.command, universal_newlines=True, env=dict(os.environ, **exportEnviron), shell=True, stderr=subprocess.STDOUT))

    def inDMC(self):
        return self.inDMC
    
    def parse_mapfile(self, jobid, mapfile, case):
        import numpy as np
        """ parse_mapfile(jobid, mapfile, case)
        """
        if (self.inDMC):
            keywords=['mapfile','mapQfile','mapUfile']
        else:
            keywords=['mapfile']
        if (case==2): keywords = [        k+'2' for k in keywords]
        if (case==3): keywords = ['extra'+k     for k in keywords]
        if (case==4): keywords = ['extra'+k+'2' for k in keywords]
    
        ##print case, mapfile,type(mapfile)
        if isinstance(mapfile, (list,tuple)): # list
            if (isinstance(mapfile[0], (list,tuple))  and (case==1 or case==2) ): # list of list
                ##print '==>List of list'
                if (self.inDMC):
                    if (case==1):kw = ["listmapweights1_",["listmapfiles1_","listmapfilesQ1_","listmapfilesU1_"]]
                    if (case==2):kw = ["listmapweights2_",["listmapfiles2_","listmapfilesQ2_","listmapfilesU2_"]]
                else:
                    if (case==1):kw = ["listmapweights1_",["listmapfiles1_"]]
                    if (case==2):kw = ["listmapweights2_",["listmapfiles2_"]]
                for i in range(len(mapfile)):
                    si=str(i+1)
                    if isinstance(mapfile[i][0],(type(1),type(1.),np.float32,np.float64)):
                        self.mySet(jobid, "%s%s"%(kw[0],si), mapfile[i][0])
                    else:
                        print ('Warning: number was expected, got '+str(mapfile[i][0]))
                    if isinstance(mapfile[i][1], (list,tuple)):
                        nl = len(mapfile[i][1])
                        if (nl != 1 and nl != 3):
                            print ('Warning: expected a 1 or 3-element list, found '+str(nl))
                            print (mapfile[i][1])
                        for j in range(nl):
                            self.mySet(jobid, "%s%s"%(kw[1][j],si), mapfile[i][1][j])
                    else:
                        self.mySet(jobid, "%s%s"%(kw[1][j],si), mapfile[i][1])
            else: # simple list
                ##print '==>List'
                if (len(mapfile[0])>0): # non-empty string in list
                    self.mySet(jobid, keywords[0], mapfile[0])
                    if (len(mapfile) == 3):# 3 element list
                        self.mySet(jobid, keywords[1], mapfile[1])
                        self.mySet(jobid, keywords[2], mapfile[2])
        else: #  not list -> string
            ##print '==>String'
            if (len(mapfile)>0): # non-empty string
                self.mySet(jobid, keywords[0],     mapfile)
    

# =======================================================================================
            
def ispice(mapin1, clout, nlmax=-1, \
           alm1fileout="", \
           alm2fileout="", \
           apodizetype=0, \
           apodizesigma="NO", \
           beam1="NO", beam2="NO", \
           beam_file1="", beam_file2="", \
           binpath="", \
           cl_inmap_file="", \
           cl_inmask_file="", \
           cl_outmap_file="", \
           cl_outmask_file="", \
           corfile="", \
           covfileout="", \
           decouple="NO", \
           extramapfile1="", extramapfile2="", \
           fits_out="YES", \
           kernelsfileout="", \
           mapfile2="", \
           maskfile1="", maskfile2="", maskfilep1="", maskfilep2="", \
           noiseclfile="", \
           noisecorfile="",\
           normfac=1.0,     \
           omp_num_threads=ompntdef, \
           pixelfile="YES", pixelfile2="YES", polarization="NO", \
           subav="NO", \
           subdipole="NO", \
           symmetric_cl="NO", \
           tenormfilein="",tenormfileout="", \
           tf_file="", \
           thetamax="NO" ,\
           tolerance="NO", \
           weightfile1="", weightfilep1="", \
           weightfile2="", weightfilep2="", \
           weightpower1=1.0, weightpower2=1.0, weightpowerp1=1.0, weightpowerp2=1.0, \
           windowfilein="",windowfileout="", \
           label="spice", submit=None):

    mytools = tools(mapin1)

    command="spice"
    if (not mytools.inDMC and len(binpath)>0):
        command=binpath
    myJob = mytools.myNewJob(command, label = label, cast = True)
    
    if (len(alm1fileout)>0):
        mytools.mySet(myJob, "alm1fileout", alm1fileout)
    if (len(alm2fileout)>0):
        mytools.mySet(myJob, "alm2fileout", alm2fileout)

    mytools.mySet(myJob, "apodizesigma", apodizesigma)
    mytools.mySet(myJob, "apodizetype",  apodizetype)
    
    mytools.mySet(myJob, "beam",         beam1)
    if (len(beam_file1)>0):
        mytools.mySet(myJob, "beam_file",    beam_file1)
    mytools.mySet(myJob, "beam2",        beam2)
    if (len(beam_file2)>0):
        mytools.mySet(myJob, "beam_file2",   beam_file2)
        
    mytools.mySet(myJob, "clfile",       clout)

    if (len(cl_inmap_file)>0):
        mytools.mySet(myJob, "cl_inmap_file",  cl_inmap_file)
    if (len(cl_inmask_file)>0):
        mytools.mySet(myJob, "cl_inmask_file",  cl_inmask_file)
    if (len(cl_outmap_file)>0):
        mytools.mySet(myJob, "cl_outmap_file",  cl_outmap_file)
    if (len(cl_outmask_file)>0):
        mytools.mySet(myJob, "cl_outmask_file",  cl_outmask_file)

    if (len(covfileout)>0):
        mytools.mySet(myJob, "covfileout",   covfileout)
    if (len(corfile)>0):
        mytools.mySet(myJob, "corfile",  corfile)
    mytools.mySet(myJob, "decouple",     decouple)

    mytools.parse_mapfile(myJob, mapin1,        1)
    mytools.parse_mapfile(myJob, mapfile2,      2)
    mytools.parse_mapfile(myJob, extramapfile1, 3)
    mytools.parse_mapfile(myJob, extramapfile2, 4)

    if (not mytools.inDMC):
        mytools.mySet(myJob, "fits_out", fits_out)
        
    if (len(kernelsfileout)>0):
        mytools.mySet(myJob, "kernelsfileout", kernelsfileout)

    if (len(maskfile1)>0):
        mytools.mySet(myJob, "maskfile",      maskfile1)
    if (len(maskfilep1)>0):
        mytools.mySet(myJob, "maskfilep",     maskfilep1)
    if (len(maskfile2)>0):
        mytools.mySet(myJob, "maskfile2",     maskfile2)
    if (len(maskfilep2)>0):
        mytools.mySet(myJob, "maskfilep2",    maskfilep2)
        
    if (len(weightfile1)>0):
        mytools.mySet(myJob, "weightfile",      weightfile1)
    if (len(weightfilep1)>0):
        mytools.mySet(myJob, "weightfilep",     weightfilep1)
    if (len(weightfile2)>0):
        mytools.mySet(myJob, "weightfile2",     weightfile2)
    if (len(weightfilep2)>0):
        mytools.mySet(myJob, "weightfilep2",    weightfilep2)
        
    mytools.mySet(myJob, "nlmax",        nlmax)
    if (len(noiseclfile)>0):
        mytools.mySet(myJob, "noiseclfile",    noiseclfile)
    if (len(noisecorfile)>0):
        mytools.mySet(myJob, "noisecorfile",    noisecorfile)

    #mytools.mySet(myJob, "normfac",      "1.00000")
    mytools.mySet(myJob, "normfac",   normfac)
    mytools.mySet(myJob, "npairsthreshold", "0.00000")
    mytools.mySet(myJob, "overwrite",       "YES")
    mytools.mySet(myJob, "polarization",    polarization)
    mytools.mySet(myJob, "pixelfile",       pixelfile)
    mytools.mySet(myJob, "pixelfile2",      pixelfile2)
    mytools.mySet(myJob, "subav",           subav)
    mytools.mySet(myJob, "subdipole",       subdipole)
    mytools.mySet(myJob, "symmetric_cl",    symmetric_cl)
    if (len(tenormfilein)>0):
        mytools.mySet(myJob, "tenormfilein", tenormfilein)
    if (len(tenormfileout)>0):
        mytools.mySet(myJob, "tenormfileout",tenormfileout)
    if (len(tf_file)>0):
        mytools.mySet(myJob, "tf_file",     tf_file)
    mytools.mySet(myJob, "thetamax",        thetamax)
    mytools.mySet(myJob, "tolerance",       tolerance)
    mytools.mySet(myJob, "verbosity",       "2")
    mytools.mySet(myJob, "weightpower",     weightpower1)
    mytools.mySet(myJob, "weightpower2",    weightpower2)
    mytools.mySet(myJob, "weightpowerp",    weightpowerp1)
    mytools.mySet(myJob, "weightpowerp2",   weightpowerp2)
    if (len(windowfilein)>0):
        mytools.mySet(myJob, "windowfilein",    windowfilein)
    if (len(windowfileout)>0):
        mytools.mySet(myJob, "windowfileout",   windowfileout)

    if (mytools.inDMC):
        mytools.mySet(myJob, "pbs_extraOption","-l place=scatter:excl -l select=1:ncpus="+str(omp_num_threads))


    mysubmit=submit
    if (mysubmit==None):
        if (mytools.inDMC):
            mysubmit=False # default: do not submit in DMC
        else:
            mysubmit=True # default: do submit in FITS

    if mysubmit:
        exev = None
        if str(omp_num_threads)!="0": exev={"OMP_NUM_THREADS":str(omp_num_threads)}
        mytools.mySubmit(myJob, exportEnviron=exev)
    else:
        if (not mytools.inDMC):
            print (mytools.command)
            print 
            print ('The command above was NOT RUN.')
            print ('Select submit=True in ispice to run it, or copy and paste it in a terminal')
    return myJob

ispice.__doc__ =     """
        ispice(mapin1, clout, nlmax=-1,
           alm1fileout="",
           alm2fileout="",
           apodizetype=0, 
           apodizesigma="NO", 
           beam1="NO", beam2="NO", 
           beam_file1="", beam_file2="",
           binpath="",
           cl_inmap_file="", 
           cl_inmask_file="", 
           cl_outmap_file="", 
           cl_outmask_file="", 
           covfileout="", 
           corfile="", 
           decouple="NO", 
           extramapfile1="", extramapfile2="", 
           fits_out="YES", 
           kernelsfileout="", 
           mapfile2="", 
           maskfile1="", maskfile2="", maskfilep1="", maskfilep2="",
           noiseclfile="", 
           noisecorfile="", 
           normfac="", 
           omp_num_threads={0},
           pixelfile="YES", pixelfile2="YES", polarization="NO", 
           subav="NO", 
           subdipole="NO", 
           symmetric_cl="NO",
           tenormfilein="", tenormfileout="",
           tf_file="", 
           thetamax="NO" ,
           tolerance="NO" ,
           weightfile1="", weightfilep1="", 
           weightfile2="", weightfilep2="", 
           weightpower1=1.0, weightpower2=1.0, weightpowerp1=1.0, weightpowerp2=1.0, 
           windowfilein="",windowfileout="", 
           label="spice", submit=None):

           Python2 and Python3 interface to F90 spice code

           Required:
             mapin1:  input I map, or list of [I,Q,U] maps,  DMC objects of type MAP
                      input I or IQU map,                    FITS files
                     or list of lists for on-the-fly weighted linear combination (LC) of maps:
                     eg:  [ [w1, I1 ],       [w2,  I2 ],       [w3,  I3 ],      ... ]
                     or   [ [w1,[I1]],       [w2, [I2]],       [w3, [I3]],      ... ]
                     or   [ [w1,[IQU1]],     [w2, [IQU2]],     [w3, [IQU3]]     ... ] (FITS only)
                     or   [ [w1,[I1,Q1,U1]], [w2, [I2,Q2,U2]], [w3, [I3,Q3,U3]] ... ] (DMC only)
                     so that mapin1_I = w1*I1 + w2*I2 + w3*I3 + ...
                       (and  mapin1_Q = w1*Q1 + w2*Q2 + w3*Q3 + ...)
                     where w* is a scalar number and I*,Q*,U* are MAPtype objects
                     
             clout:    output C(l) (either auto or cross),   object  of type CL
           
           Optional:
             alm1fileout:   see [1] below
             alm2fileout:   see [1] 
             apodizesigma:  see [1]
             apodizetype:   see [1]
             beam1, beam2: beam(s) FWHM in arcmin
             beam_file*: input B(l) file(s),                 FITS files or objects of type CL
             binpath: name of Spice executable, or its relative or absolute path.
                If set to 'spice' (=default value), 'spice_SP' or 'Spice_DP',
                 a code of the same name will be looked-for in the directory ({1})
                 containing this python script or its siblings.
                Paths starting with '.' are relative to current directory.
                Those starting with '/' are absolute.
             cl_*_file:     see [1]
             covfileout: output C(l)C(l) covariance matrix,  FITS file  or object  of type TAB3D
             corfile:       see [1]
             decouple:      see [1]
             extramapfile1: I or [I,Q,U] map to be added to mapin1
                will be ignored by Spice if mapin1 is a LC
             extramapfile2: I or [I,Q,U] map to be added to mapfile2
                will be ignored by Spice if mapfile2 is a LC
             fits_out: output files are in FITS instead of plain ASCII
             kernelsfileout: see [1]
             label: name given to job
             mapfile2: map(s) or LC of maps for cross-spectrum
                see mapin1 for format
             maskfile*:     see [1]
             noiseclfile:   see [1] 
             noisecorfile:  see [1] 
             normfac:       see [1] 
             omp_num_threads: number of OMP threads to be used (temporarily (re)defining OMP_NUM_THREADS if >0)
                              default value: current value of $OMP_NUM_THREADS ({0}) or 0
             pixelfile*:    see [1]
             polarization:  see [1]
             subav:         see [1]
             subdipole:     see [1]
             submit:  if set to False, the command string is generated, but job is not submitted
             symmetric_cl:  see [1]
             tenormfile*:   see [1]
             tf_file:       see [1]
             tolerane:      see [1]
             weightfile*:   see [1]
             weightpower*:  see [1]
             windowfile*:   see [1]
           
           [1] ispice.readme()
               or
               {2}

           Note: values "YES" and "NO" can be replaced by True and False respectively,
               wherever applicable.


           Examples:
           
           import ispice
           
           # compute polarized spectrum of mapfile:
           ispice.ispice(mapfile, clfile, polarization=True, pixelfile=True, subdipole=True)
           
           # help on ispice (python interface)
           help(ispice.ispice)

           # help on PolSpice (F90 backend):
           ispice.readme()
           
           
    """.format(str(ompntdef),dirname(__file__),readmehtml)
